klippy/extras/save_variables.py

```diff
@@ -0,0 +1,62 @@
+# Get and Save arbitrary variables in the config file so that values can be kept
+# across restarts.
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import sys,os,logging, ConfigParser
+
+error = ConfigParser.Error
+
+class SaveVariables:
+    error = ConfigParser.Error
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.name = config.get_name()
+        self.cfilename = config.get('filename')
+        self.filename = os.path.join(os.path.expanduser("~"), self.cfilename)
+        self.allVariables = {}
+        self.loadVariables()
+        logging.info(str(self.allVariables))
+        self.gcode.register_command(
+            'SAVE_VARIABLE', self.cmd_SAVE_VARIABLE,
+            desc=self.cmd_SAVE_VARIABLE_help)
+    cmd_SAVE_VARIABLE_help = "Save arbitrary variables to disk"
+    def loadVariables(self):
+        self.allVariables = {}
+        try:
+            self.variablefile = ConfigParser.ConfigParser()
+            self.variablefile.read(self.filename)
+            self.allVariables = dict(self.variablefile.items('Variables'))
+        except error as e:
+            msg = e.message + "\nUnable to parse existing variable file"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
+    def cmd_SAVE_VARIABLE(self, params):
+        self.variable_name = self.gcode.get_str('VARIABLE', params)
+        self.variable_value = self.gcode.get_str('VALUE', params)
+        try:
+            self.variablefile = ConfigParser.ConfigParser()
+            self.variablefile.read(self.filename)
+        except error as e:
+            msg = "Unable to parse existing variable file"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
+        try:
+            self.variablefile.set('Variables',self.variable_name,
+                              self.variable_value)
+            self.variablefile.write(open(self.filename, "w"))
+        except error as e:
+            msg = e.message + "\nUnable to save variable"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
+        self.gcode.respond_info("Variable Saved")
+        self.loadVariables()
+        logging.info(str(self.allVariables))
+    def get_status(self, eventtime):
+        #return {'temp1':self.filename}
+        #logging.info("allVariables='%s'", repr(self.allVariables))
+        return self.allVariables
```

Comment:
I'd recommend doing something like `return {"variables": self.allVariables}` as then it becomes possible to add additional exports in the future without fear of conflicts.

Reply:
How would this work in terms of the printer variable - would `printer.save_variables.currentextruder` become `printer.save_variables.variables.currentextruder` ? I'd prefer not to do that in that case. I doubt I'd add more feature to this module in any case - it's simple and works. Which I think is great.

Reply:
I've made the change.